src/compiler/checker.ts

```diff
@@ -5428,15 +5473,234 @@ namespace ts {
-                (<InterfaceTypeWithDeclaredMembers>type).declaredConstructSignatures = getSignaturesOfSymbol(symbol.members.get(InternalSymbolName.New));
+                const members = getMembersOfSymbol(symbol);
+                (<InterfaceTypeWithDeclaredMembers>type).declaredProperties = getNamedMembers(members);
+                (<InterfaceTypeWithDeclaredMembers>type).declaredCallSignatures = getSignaturesOfSymbol(members.get(InternalSymbolName.Call));
+                (<InterfaceTypeWithDeclaredMembers>type).declaredConstructSignatures = getSignaturesOfSymbol(members.get(InternalSymbolName.New));
                 (<InterfaceTypeWithDeclaredMembers>type).declaredStringIndexInfo = getIndexInfoOfSymbol(symbol, IndexKind.String);
                 (<InterfaceTypeWithDeclaredMembers>type).declaredNumberIndexInfo = getIndexInfoOfSymbol(symbol, IndexKind.Number);
             }
             return <InterfaceTypeWithDeclaredMembers>type;
         }
 
+        /**
+         * Indicates whether a type can be used as a late-bound name.
+         */
+        function isTypeUsableAsLateBoundName(type: Type): type is LiteralType | UniqueESSymbolType {
+            return !!(type.flags & TypeFlags.StringOrNumberLiteralOrUnique);
+        }
+
+        /**
+         * Indicates whether a declaration name is definitely late-bindable.
+         * A declaration name is only late-bindable if:
+         * - It is a `ComputedPropertyName`.
+         * - Its expression is an `Identifier` or either a `PropertyAccessExpression` an
+         * `ElementAccessExpression` consisting only of these same three types of nodes.
+         * - The type of its expression is a string or numeric literal type, or is a `unique symbol` type.
+         */
+        function isLateBindableName(node: DeclarationName): node is LateBoundName {
+            return isComputedPropertyName(node)
+                && isEntityNameExpression(node.expression)
+                && isTypeUsableAsLateBoundName(checkComputedPropertyName(node));
+        }
+
+        /**
+         * Indicates whether a declaration has a late-bindable dynamic name.
+         */
+        function hasLateBindableName(node: Declaration): node is LateBoundDeclaration {
+            const name = getNameOfDeclaration(node);
+            return name && isLateBindableName(name);
+        }
+
+        /**
+         * Indicates whether a declaration has a dynamic name that cannot be late-bound.
+         */
+        function hasNonBindableDynamicName(node: Declaration) {
+            return hasDynamicName(node) && !hasLateBindableName(node);
+        }
+
+        /**
+         * Indicates whether a declaration name is a dynamic name that cannot be late-bound.
+         */
+        function isNonBindableDynamicName(node: DeclarationName) {
+            return isDynamicName(node) && !isLateBindableName(node);

```

Comment:
It turns out that will be a slight issue going forward, as the definitions for NodeJS also define `Symbol.iterator` as `symbol`. We run into a bit of a versioning issue if we update our default lib to use `unique symbol` as they will collide with the NodeJS definition. While we can version the NodeJS definition against typescript, we only do that for shipping versions. 

Since we are not replacing the "well known symbol" concept in this PR, its not an issue I plan to address here.